src/main/kotlin/org/rust/ide/template/postfix/stringBasedPostfixTemplates.kt

```diff
@@ -0,0 +1,136 @@
+import org.rust.lang.core.resolve.RustResolveEngine
+import org.rust.lang.core.resolve.scope.RustResolveScope
+import org.rust.lang.core.types.RustEnumType
+import org.rust.lang.core.types.util.resolvedType
+
+class AssertPostfixTemplate : StringBasedPostfixTemplate(
+    "assert",
+    "assert!(expr);",
+    RustTopMostInScopeSelector(RustExprElement::isBool)
+) {
+    override fun getTemplateString(element: PsiElement): String? =
+        "assert!(\$expr$);\$END$"
+
+    override fun getElementToRemove(expr: PsiElement): PsiElement = expr
+}
+
+class AssertEqPostfixTemplate : StringBasedPostfixTemplate(
+    "asserteq",
+    "assert_eq!(lhs, rhs);",
+    RustTopMostInScopeSelector(RustExprElement::isEqExpr)
+) {
+    override fun getTemplateString(element: PsiElement): String? {
+        if (element !is RustBinaryExprElement) return null
+        return "assert_eq!(${element.left.text}, ${element.right?.text});\$END$"
+    }
+
+    override fun getElementToRemove(expr: PsiElement): PsiElement = expr
+}
+
+class DebugAssertPostfixTemplate : StringBasedPostfixTemplate(
+    "assertdeb",
+    "debug_assert!(expr);",
+    RustTopMostInScopeSelector(RustExprElement::isBool)
+) {
+    override fun getTemplateString(element: PsiElement): String? =
+        "debug_assert!(\$expr$);\$END$"
+
+    override fun getElementToRemove(expr: PsiElement): PsiElement = expr
+}
+
+class DebugAssertEqPostfixTemplate : StringBasedPostfixTemplate(
+    "assertdebeq",
+    "debug_assert_eq!(lhs, rhs);",
+    RustTopMostInScopeSelector(RustExprElement::isEqExpr)
+) {
+    override fun getTemplateString(element: PsiElement): String? {
+        if (element !is RustBinaryExprElement) return null
+        return "debug_assert_eq!(${element.left.text}, ${element.right?.text});\$END$"
+    }
+
+    override fun getElementToRemove(expr: PsiElement): PsiElement = expr
+}
+
+class ParenthesesPostfixTemplate : StringBasedPostfixTemplate(
+    "par",
+    "(expr)",
+    RustAllParentsInScopeSelector(RustExprElement::all)
+) {
+    override fun getTemplateString(element: PsiElement): String? =
+        "(\$expr$)\$END$"
+
+    override fun getElementToRemove(expr: PsiElement): PsiElement = expr
+}
+
+class MatchPostfixTemplate : StringBasedPostfixTemplate(
+    "match",
+    "match expr {...}",
+    RustTopMostInScopeSelector(RustExprElement::isEnum)
+) {
+    override fun getTemplateString(element: PsiElement): String? {
+        val enumType = (element as RustExprElement).resolvedType as RustEnumType
+
+        val allAvailableNames = generateSequence(
+            element.parentOfType<RustResolveScope>(),
+            { it.parentOfType<RustResolveScope>() }
+        )
+            .flatMap { RustResolveEngine.declarations(it, element) }
+            .filter { it.element != null }
+            .map { RustPsiImplUtil.canonicalCratePath(it.element!!) }
+            .filterNotNull()
+            .map { it.toString() }
+            .toSet()
```

Comment:
This can be simplified a bit.

`filter { it.element != null}` => `mapNotNull { it.element }`

`map { ... }. filterNotNull()` => `mapNotNull { ... }`